### PR TITLE
hal/ia32/console-vga: change colors using ANSI escape sequences

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

RTOS-140

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by agkaminski — APPROVED** [review]
LGTM, let's wait for @Maxez approve

File: hal/ia32/console.c [hunk 1]
```diff
@@ -21,133 +21,343 @@
...
+	/* Initialize VGA */
+	halconsole_common.vram = (u16 *)(VADDR_KERNEL + (color ? 0xb8000 : 0xb0000));
+	halconsole_common.crtc = (void *)(color ? 0x3d4 : 0x3b4);
 
-	hal_spinlockCreate(&console.spinlock, "console.spinlock");
-	return;
+	/* Default 80x25 text mode with magenta color attribute */
+	halconsole_common.rows = 25;
+	halconsole_common.cols = 80;
+	halconsole_common.attr = 0x05;
```
Comment (Maxez) [inline hal/ia32/console.c:361]: In kernel we want to initialize default color attribute to cyan - I'd change the comment and set `halconsole_common.attr = 0x03`

Comment (HBuczynski) [reply hal/ia32/console.c]: Done

**Review by Maxez — APPROVED** [review]
LGTM

File: hal/ia32/console.c [hunk 2]
```diff
@@ -21,133 +21,343 @@
...
+
+
+static void console_memset(volatile u16 *vram, u16 val, unsigned int n)
+{
+	unsigned int i;
 
-	int attr;
-} console;
+	for (i = 0; i < n; i++)
+		*(vram + i) = val;
```
Comment (kemonats) [inline hal/ia32/console.c:65]: Maybe inline assembler and using `movs`. It will be faster.

File: hal/ia32/console.c [hunk 3]
```diff
@@ -21,133 +21,343 @@
...
-			break;
+							row = halconsole_common.parms[0] - 1;
+							col = halconsole_common.parms[1] - 1;
+							halconsole_common.esc = esc_init;
+							break;
+
+						case 'J':
+							switch (halconsole_common.parms[0]) {
+								case 0:
+									console_memset(halconsole_common.vram + row * halconsole_common.cols + col, (u16)halconsole_common.attr << 8 | ' ', halconsole_common.cols * (halconsole_common.rows - row) - col);
```
Comment (kemonats) [inline hal/ia32/console.c:195]: A line that is 250 characters long?

File: hal/ia32/console.c [hunk 4]
```diff
@@ -21,133 +21,343 @@
...
+									break;
+
+								case 1:
+									console_memset(halconsole_common.vram, (u16)halconsole_common.attr << 8 | ' ', row * halconsole_common.cols + col + 1);
+									break;
+
+								case 2:
+									console_memset(halconsole_common.vram, (u16)halconsole_common.attr << 8 | ' ', halconsole_common.rows * halconsole_common.cols);
+									break;
+							}
```
Comment (kemonats) [inline hal/ia32/console.c:205]: A line that is 250 characters long?
Maybe someting like that.
```suggestion
							; /* This empty statement must be */
							volatile u16 *vram = halconsole_common.vram;
							const u16 val = (u16)halconsole_common.attr << 8 | ' ';
							unsigned int n = 0;

							switch (halconsole_common.parms[0]) {
								case 0:
									vram += row * halconsole_common.cols + col;
									n = halconsole_common.cols * (halconsole_common.rows - row) - col;
									break;

								case 1:
									// vram = halconsole_common.vram;
									n = row * halconsole_common.cols + col + 1;
									break;

								case 2:
									// vram = halconsole_common.vram;
									n = halconsole_common.rows * halconsole_common.cols;
									break;
							}
							if (n > 0)
								console_memset(vram, val, n);

```

File: hal/ia32/console.c (hunk 2 above)
Comment (HBuczynski) [reply hal/ia32/console.c:65]: This code is derived from [plo](https://github.com/phoenix-rtos/plo/blob/f5401788f72d58658a2aa1879ccfabc475cc900b/hal/ia32/console-vga.c#L79-L314) so in this PR I would not like to add significant modifications.

File: hal/ia32/console.c (hunk 4 above)
Comment (HBuczynski) [reply hal/ia32/console.c:205]: This code is derived from [plo](https://github.com/phoenix-rtos/plo/blob/f5401788f72d58658a2aa1879ccfabc475cc900b/hal/ia32/console-vga.c#L79-L314) so it slightly differs from master branch.

`volatile` is correct, the vga hardware uses simultaneously from `halconsole_common.vram`.

We prefer not to write variables declaration inside `switch ... case`. I leave it as it is.

Comment (kemonats) [reply hal/ia32/console.c:205]: But it seems to me that the hardware only reads from vram, it doesn't change its contents. You're just changing the contents of that memory.

Comment (agkaminski) [reply hal/ia32/console.c:205]: If hardware only reads this memory it must be volatile anyway - we need to make sure that we actually write to this memory

Comment (kemonats) [reply hal/ia32/console.c:205]: If that's the case, what's changed that isn't in the master branch but is here.

**Pull request merged**